Allow NeoMutt to show complete MIME structure

rayfordshire opened the pull request:
The attachment menu hides some MIME parts from the user which are usually uninteresting, e.g. a multipart/mixed MIME part at the top level as it is only the container for the attachments.

However, when debugging email problems, especially when encryption is involved, it is useful to see the complete MIME structure. Example where this would have been useful is https://github.com/neomutt/neomutt/issues/3711 I also had a couple of incidents in the past, where I would have liked to have this feature instead of resorting back to an editor and openssl.

Introduce a new config variable $attach_complete_mime_structure, which defaults to false.  If this variable is set, then NeoMutt does not hide any MIME parts from the user.  If unset, it does what it currently does and hides some MIME parts from the user.

---

@gahr I'm sorry, but yes this is a new config variable.  But I don't see another way if the default should be "hide uninteresting stuff from people".

---

* **What are the relevant issue numbers?**

Related: https://github.com/neomutt/neomutt/issues/3711
Fixes: https://github.com/neomutt/neomutt/issues/3706

gahr commented:
I wonder if we shouldn't do it always.

flatcap commented:
> I wonder if we shouldn't do it always.

I'm in favour of that, too.

Compose isn't consistent with received mail -- it's not always fully-formed.
When the email is sent, `mutt_send_message()` checks and adds a wrapper if necessary.

https://github.com/neomutt/neomutt/blob/fd4e14617f95503a36a6bb3162bbe25e56ac3b77/send/send.c#L2817-L2818

rayfordshire commented:
Reworked the initial proposal, the commits are now:

1.  removing the check to hide some MIME-parts
2. / 3. Two mini improvements
3.-end Attempt to show MIME part in compose menu

The last part is currently [WIP]; the main problem is to get a fully parsed `Body` (or `Email`).  I've modified `mutt_make_message_attach()` which covers the "add mail as attachment" case.  However, postponing still needs to be handled.

**Thus, the first question** when/where are emails usually parsed, so that a `Body` represents the complete MIME structure, i.e. `body->parts` is only NULL if the message does not contain any MIME-parts.  I could add those parsing step for postpone in `postpone/postpone.c` function `mutt_get_postponed()` somewhere around here:

(A hint which function to call would be greatly appreciated)

https://github.com/neomutt/neomutt/blob/b89d8f3f27045d720b5017f6b8ff8bfb18857392/postpone/postpone.c#L714-L719

Another problem is when viewing those MIME-subparts is that they are not backed by a file.  I image the same behaviour as with the regular attachment menu. I tried setting the filename but that a) didn't work and b) made the attachment list look ugly.

**Second question**: How can the MIME-subparts be shown (with function `view-attach`).

flatcap commented:
Rebased over `main` (NeoMutt 2023-03-22).

rayfordshire commented:
How do we proceed here?

The first three commits fix the problem for the attach menu (and can be merged), the others are concerned with the compose menu and have open questions: https://github.com/neomutt/neomutt/pull/3738#issuecomment-1453315196

flatcap commented:
> The first three commits fix the problem for the attach menu (and can be merged)

Thanks.  Done.

flatcap commented:
Archived to: https://github.com/neomutt/neomutt-old/tree/show-complete-mime-structure
